### PR TITLE
Add clang-16

### LINKED ISSUE #248: Clang wrapper for Windows isn't working
These are some of the issues of our Clang wrapper for Windows:
```console
% julia --compile=min -e 'using BinaryBuilderBase; BinaryBuilderBase.runshell(Platform("x86_64", "windows"))'                          
sandbox:${WORKSPACE} # echo 'int main(){return 0;}' | clang -x c -
clang-13: warning: argument unused during compilation: '--gcc-toolchain=/opt/x86_64-w64-mingw32' [-Wunused-command-line-argument]
clang-13: warning: argument unused during compilation: '-stdlib=libstdc++' [-Wunused-command-line-argument]
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc_eh
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc
/opt/x86_64-w64-mingw32/bin/x86_64-w64-mingw32-ld: cannot find -lgcc_eh
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
sandbox:${WORKSPACE} # echo -e '#include <limits>\nint main(){return 0;}' | clang -x c++ -
clang-13: warning: argument unused during compilation: '--gcc-toolchain=/opt/x86_64-w64-mingw32' [-Wunused-command-line-argument]
<stdin>:1:10: fatal error: 'limits' file not found
#include <limits>
         ^~~~~~~~
1 error generated.
```
We actually already do https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/4b21cde2105830e287271c323d909b9ee8a64c5e/src/Runner.jl#L318-L319 to make Clang find the GCC toolchain, but it doesn't seem to be effective on Windows, maybe Clang expects a different tree structure, or maybe it's just ignored (why?), given the warning about the argument being unused.

I tried to address this issue in #247, but it's a horrible ad-hoc kludge.  I opened this issue because I feel like we should be able to make this work more nicely than putting a bunch of random `-I...` and `-L...` arguments.

### LINKED ISSUE #290: Clang wrapper for Linux and Windows should set `_GLIBCXX_USE_CXX11_ABI`
At the moment https://github.com/JuliaPackaging/BinaryBuilderBase.jl/blob/f782dc5d327dfa21cc26ae582e4d6424f7df8bc5/src/Runner.jl#L462-L467 is done only for GCC, but it should be done also for Clang on Linux and Windows, since in that case we use libstdc+.

### PR DESCRIPTION
**Pull request by gbaraldi**
This fixes #290 and fixes #248

### CODE REVIEW

File: src/Runner.jl [hunk 1]
```diff
@@ -415,6 +421,18 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
...
+        if !Sys.isbsd(p)
+            gcc_version, llvm_version = select_compiler_versions(p, compilers)
+            append!(flags, String["-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/$(aatriplet(p))",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include/c++/$(gcc_version)/backward",
+                                  "-I/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)/include",
+                                  "-I/opt/$(aatriplet(p))/lib/gcc/$(aatriplet(p))/$(gcc_version)/include-fixed",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/include",
+                                  "-I/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root/include"])
+        end
```
Comment (giordano) [inline src/Runner.jl:434]: Why this instead of `--gcc-toolchain`?  Did you see #247 and #248?  One problem of this approach is also that this triggers `-Werror` failures in case some paths are unused, which isn't great.

**Review by giordano — COMMENTED** [review]
It'd be good to see the corresponding PR to Yggdrasil as well.

File: src/Runner.jl [hunk 2]
```diff
@@ -322,17 +322,23 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
...
         if !Sys.isbsd(p)
-            append!(flags, [
-                # Find GCC toolchain here (for things like libgcc_s)
-                "--gcc-toolchain=/opt/$(aatriplet(p))"
-            ])
+            if cxxstring_abi(p) == "cxx11"
+                push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=1")
+            elseif cxxstring_abi(p) == "cxx03"
+                push!(flags, "-D_GLIBCXX_USE_CXX11_ABI=0")
+            end
```
Comment (giordano) [inline src/Runner.jl:329]: Maybe define a function to be used both in GCC and Clang, to avoid duplication?

File: src/Runner.jl (hunk 1 above)
Comment (gbaraldi) [reply src/Runner.jl:434]: gcc-toolchain wasn't reliable for me when testing, our gcc install is a bit weird because we set a sysroot that isn't the usual one and clang isn't able to find stuff. Specially in mingw.
About the warning, given that it didn't complain when compiling very simple C code I would be surprised if it did for more complicated ones?

Comment (giordano) [reply src/Runner.jl:434]: I'm pretty sure I've seen that warning in the wild.

Comment (gbaraldi) [reply src/Runner.jl:434]: I discovered why it might complain, we should use -isystem here. Those if unused do not get warnings.

Comment (giordano) [reply src/Runner.jl:434]: Ah, that's probably a good idea here

**Comment by gbaraldi** [comment]
Given that this successfully built LLVM I think we can say it's working

**Comment by giordano** [comment]
Thanks!

File: src/Runner.jl [hunk 3]
```diff
@@ -322,17 +332,17 @@ function generate_compiler_wrappers!(platform::AbstractPlatform; bin_path::Abstr
             "--sysroot=/opt/$(aatriplet(p))/$(aatriplet(p))/sys-root",
         ])
         if !Sys.isbsd(p)
-            append!(flags, [
-                # Find GCC toolchain here (for things like libgcc_s)
-                "--gcc-toolchain=/opt/$(aatriplet(p))"
-            ])
```
Comment (giordano) [inline src/Runner.jl:328]: Why was this removed?
